File: win-prog/ProgWinVB/Toolbars and Status Bars/ToggleButtons/ToggleButtons.vb
'----------------------------------------------
' ToggleButtons.vb (c) 2002 by Charles Petzold
'----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class ToggleButtons
    Inherits Form

    Protected pnl As Panel
    Protected tbar As ToolBar
    Protected strText As String = "Toggle"
    Protected clrText As Color = SystemColors.WindowText
    Private fntstyle As FontStyle = FontStyle.Regular

    Shared Sub Main()
        Application.Run(New ToggleButtons())
    End Sub

    Sub New()
        Text = "Toggle Buttons"

        ' Create panel to fill interior.
        pnl = New Panel()
        pnl.Parent = Me
        pnl.Dock = DockStyle.Fill
        pnl.BackColor = SystemColors.Window
        pnl.ForeColor = SystemColors.WindowText
        AddHandler pnl.Resize, AddressOf PanelOnResize
        AddHandler pnl.Paint, AddressOf PanelOnPaint

        ' Create ImageList.
        Dim bm As New Bitmap(Me.GetType(), "FontStyleButtons.bmp")
        Dim imglst As New ImageList()
        imglst.ImageSize = New Size(bm.Width \ 4, bm.Height)
        imglst.Images.AddStrip(bm)
        imglst.TransparentColor = Color.White

        ' Create ToolBar.
        tbar = New ToolBar()
        tbar.ImageList = imglst
        tbar.Parent = Me
        tbar.ShowToolTips = True
        AddHandler tbar.ButtonClick, AddressOf ToolBarOnClick

        ' Create ToolBarButtons.
        Dim afs() As FontStyle = {FontStyle.Bold, FontStyle.Italic, _
                                  FontStyle.Underline, FontStyle.Strikeout}
        Dim i As Integer
        For i = 0 To 3
            Dim tbarbtn As New ToolBarButton()
            tbarbtn.ImageIndex = i
            tbarbtn.Style = ToolBarButtonStyle.ToggleButton
            tbarbtn.ToolTipText = afs(i).ToString()
            tbarbtn.Tag = afs(i)
            tbar.Buttons.Add(tbarbtn)
        Next i
    End Sub

    Private Sub ToolBarOnClick(ByVal obj As Object, _
                               ByVal tbbcea As ToolBarButtonClickEventArgs)
        Dim tbb As ToolBarButton = tbbcea.Button

        ' If the Tag isn't a FontStyle, don't do anything.
        If tbb.Tag Is Nothing Then Return
        If Not tbb.Tag.GetType() Is GetType(FontStyle) Then Return

        ' Set or clear the bit in the fntstyle field.
        If tbb.Pushed Then
            fntstyle = fntstyle Or CType(tbb.Tag, FontStyle)
        Else
            fntstyle = fntstyle And Not CType(tbb.Tag, FontStyle)
        End If
        pnl.Invalidate()
    End Sub

    Private Sub PanelOnResize(ByVal obj As Object, ByVal ea As EventArgs)
        Dim pnl As Panel = DirectCast(obj, Panel)
        pnl.Invalidate()
    End Sub

    Private Sub PanelOnPaint(ByVal obj As Object, _
                             ByVal pea As PaintEventArgs)
        Dim pnl As Panel = DirectCast(obj, Panel)
        Dim grfx As Graphics = pea.Graphics
        Dim fnt As New Font("Times New Roman", 72, fntstyle)
        Dim szf As SizeF = grfx.MeasureString(strText, fnt)

        grfx.DrawString(strText, fnt, New SolidBrush(clrText), _
                        (pnl.Width - szf.Width) / 2, _
                        (pnl.Height - szf.Height) / 2)
    End Sub
End Class

File: win-prog/ProgWinCS/Toolbars and Status Bars/ToggleButtons/ToggleButtons.cs
//--------------------------------------------
// ToggleButtons.cs © 2001 by Charles Petzold
//--------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class ToggleButtons: Form
{
     protected Panel   panel;
     protected ToolBar tbar;
     protected string  strText = "Toggle";
     protected Color   clrText = SystemColors.WindowText;
     FontStyle         fontstyle = FontStyle.Regular;

     public static void Main()
     {
          Application.Run(new ToggleButtons());
     }
     public ToggleButtons()
     {
          Text = "Toggle Buttons";

               // Create panel to fill interior.

          panel = new Panel();
          panel.Parent = this;
          panel.Dock = DockStyle.Fill;
          panel.BackColor = SystemColors.Window;
          panel.ForeColor = SystemColors.WindowText;
          panel.Resize += new EventHandler(PanelOnResize);
          panel.Paint += new PaintEventHandler(PanelOnPaint);

               // Create ImageList.

          Bitmap bm = new Bitmap(GetType(), 
                                 "ToggleButtons.FontStyleButtons.bmp");

          ImageList imglst = new ImageList();
          imglst.ImageSize = new Size(bm.Width / 4, bm.Height);
          imglst.Images.AddStrip(bm);
          imglst.TransparentColor = Color.White;

               // Create ToolBar.

          tbar = new ToolBar();
          tbar.ImageList = imglst;
          tbar.Parent = this;
          tbar.ShowToolTips = true;
          tbar.ButtonClick += 
                         new ToolBarButtonClickEventHandler(ToolBarOnClick);

               // Create ToolBarButtons.

          FontStyle[] afs = { FontStyle.Bold, FontStyle.Italic,
                              FontStyle.Underline, FontStyle.Strikeout };

          for (int i = 0; i < 4; i++)
          {
               ToolBarButton tbarbtn = new ToolBarButton();
               tbarbtn.ImageIndex = i;
               tbarbtn.Style = ToolBarButtonStyle.ToggleButton;
               tbarbtn.ToolTipText = afs[i].ToString();
               tbarbtn.Tag = afs[i];

               tbar.Buttons.Add(tbarbtn);
          }
     }
     void ToolBarOnClick(object obj, ToolBarButtonClickEventArgs tbbcea)
     {
          ToolBarButton tbarbtn = tbbcea.Button;

               // If the Tag isn't a FontStyle, don't do anything.

          if (tbarbtn.Tag == null || 
			  tbarbtn.Tag.GetType() != typeof(FontStyle))
               return;

               // Set or clear the bit in the fontstyle field.

          if (tbarbtn.Pushed)
               fontstyle |= (FontStyle) tbarbtn.Tag;
          else
               fontstyle &= ~(FontStyle) tbarbtn.Tag;

          panel.Invalidate();
     }
     void PanelOnResize(object obj, EventArgs ea)
     {
          Panel panel = (Panel) obj;
          panel.Invalidate();
     }
     void PanelOnPaint(object obj, PaintEventArgs pea)
     {
          Panel    panel = (Panel) obj;
          Graphics grfx  = pea.Graphics;
          Font     font  = new Font("Times New Roman", 72, fontstyle);
          SizeF    sizef = grfx.MeasureString(strText, font);

          grfx.DrawString(strText, font, new SolidBrush(clrText),
                          (panel.Width - sizef.Width) / 2,
                          (panel.Height - sizef.Height) / 2);
     }
}